homeassistant/components/combined_energy/sensor.py

```diff
@@ -0,0 +1,270 @@
+            raw_value = raw_value[-1]
+        return round(raw_value, self.native_value_rounding)
+
+
+class EnergySensor(CombinedEnergyReadingsSensor):
+    """Sensor for energy readings."""
+
+    @property
+    def last_reset(self) -> datetime | None:
+        """Last time the data was reset."""
+        if device_readings := self.device_readings:
+            return device_readings.range_start
+        return None
+
+    def _to_native_value(self, raw_value):
+        """Convert non-none raw value into usable sensor value."""
+        value = sum(raw_value)
+        return round(value, self.native_value_rounding)
+
+
+class PowerSensor(CombinedEnergyReadingsSensor):
+    """Sensor for power readings."""
+
+    def _to_native_value(self, raw_value):
+        """Convert non-none raw value into usable sensor value."""
+        return round(raw_value, self.native_value_rounding)
+
+
+class PowerFactorSensor(CombinedEnergyReadingsSensor):
+    """Sensor for power factor readings."""
+
+    native_value_rounding = 1
+
+    def _to_native_value(self, raw_value):
+        """Convert non-none raw value into usable sensor value."""
+        # The API expresses the power factor as a fraction convert to %
+        return round(raw_value[-1] * 100, self.native_value_rounding)
+
+
+class WaterVolumeSensor(CombinedEnergyReadingsSensor):
+    """Sensor for water volume readings."""
+
+    def _to_native_value(self, raw_value):
+        """Convert non-none raw value into usable sensor value."""
+        return int(round(raw_value[-1], 0))
+
+
+# Map of common device classes to specific sensor types
+SENSOR_TYPE_MAP: dict[
+    SensorDeviceClass | str | None, type[CombinedEnergyReadingsSensor]
+] = {
+    SensorDeviceClass.ENERGY: EnergySensor,
+    SensorDeviceClass.POWER: PowerSensor,
+    SensorDeviceClass.WATER: WaterVolumeSensor,
+    SensorDeviceClass.POWER_FACTOR: PowerFactorSensor,
+    None: GenericSensor,
+}
+
+
+class CombinedEnergyReadingsSensorFactory:
+    """
+    Factory for generating devices/entities.
+
+    Entities/Devices are described in the installation model.
+    """
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        installation: Installation,
+        readings: CombinedEnergyReadingsDataService,
+    ) -> None:
+        """Initialise readings sensor factory."""
+        self.hass = hass
+        self.installation = installation
+        self.readings = readings
+
+    def _generate_device_info(self, device: Device) -> DeviceInfo:
+        """Generate device info from API device response."""
+        return DeviceInfo(
+            identifiers={
+                (
+                    DOMAIN,
+                    f"install_{self.installation.installation_id}-device_{device.device_id}",
+                )
+            },
+            manufacturer=device.device_manufacturer,
+            model=device.device_model_name,
+            name=device.display_name,
+        )
```

Comment:
Instead of generating device info, you should be able to just create a CombinedEnergyEntity object, do it in the init there and then make all of your other entities based off of that one

Reply:
instead of calling this code, you can create the following:

```
CombinedEnergyEntity(CoordinatorEntity):

def __init__(self,device, ...):
self._attr_device_info = DeviceInfo(
            identifiers={
                (
                    DOMAIN,
                    f"install_{self.installation.installation_id}-device_{device.device_id}",
                )
            },
            manufacturer=device.device_manufacturer,
            model=device.device_model_name,
            name=device.display_name,
        )
```

Then your CombinedEnergyReadingsSensor(CombinedEnergyEntity) and you just pass it the device as you do now and then it should be handled in the super().__init__()

Reply:
The code as is creates a single `DeviceInfo` instance for each device which is then passed to each individual sensor instance.  It's a one-to-many relationship. Why create unnecessary instances of `DeviceInfo`?

Reply:
You could do it Per coordinator instead then if you'd like

Reply:
That wouldn't work either. There is a single readings co-ordinator that obtains data for all the devices related to an _installation_ when polled, not a single device. On each poll of the Readings co-ordinator the integration then fans out all of the data to the various devices and sensors. The design I have used is the simplest I could provide that correctly constructs the required objects.

There is a separate LogSession Co-ordinator (Combined Energy term) that needs to be pinged periodically to keep the _log-session_, I assume this is actually a mod-bus session alive or data stops being provided until it is called again.

Much of the complexity of this integration comes from the design of the API provided by Combined Energy.